enterprise/backend/src/metabase_enterprise/caching/strategies.clj

```diff
@@ -0,0 +1,198 @@
+   [metabase.task :as task]
+   [metabase.util.cron :as u.cron]
+   [metabase.util.log :as log]
+   [metabase.util.malli.schema :as ms]
+   [toucan2.core :as t2])
+  (:import
+   (org.quartz.spi MutableTrigger)))
+
+(set! *warn-on-reflection* true)
+
+(def CacheStrategy ;; not used in API since there are some additional fields
+  "Schema for a caching strategy"
+  (mc/schema
+   [:and
+    [:map
+     [:type [:enum :nocache :ttl :duration :schedule :query]]]
+    [:multi {:dispatch :type}
+     [:nocache  [:map
+                 [:type keyword?]]]
+     [:ttl      [:map
+                 [:type keyword?]
+                 [:multiplier ms/PositiveInt]
+                 [:min_duration ms/PositiveInt]]]
+     [:duration [:map
+                 [:type keyword?]
+                 [:duration ms/PositiveInt]
+                 [:unit [:enum "hours" "minutes" "seconds" "days"]]]]
+     [:schedule [:map
+                 [:type keyword?]
+                 [:updated_at some?]
+                 [:schedule u.cron/CronScheduleString]]]
+     [:query    [:map
+                 [:type keyword?]
+                 [:updated_at some?]
+                 [:payload [:maybe any?]]
+                 [:field_id int?]
+                 [:aggregation [:enum "max" "count"]]
+                 [:schedule string?]]]]]))
+
+(defn granular-ttl
+  "Returns the granular cache ttl (in hours) for a card. On EE, this first checking whether there is a stored value
+   for the card, dashboard, or database (in that order of decreasing preference). Returns nil on OSS."
+  [card dashboard-id]
+  ; stored TTLs are in hours; convert to seconds
+  (or (:cache_ttl card)
+      (:cache_ttl (t2/select-one [:model/Dashboard :cache_ttl] :id dashboard-id))
+      (:cache_ttl (t2/select-one [:model/Database :cache_ttl] :id (:database_id card)))))
+
+(defenterprise-schema granular-cache-strategy :- [:maybe CacheStrategy]
+  "Returns the granular cache strategy for a card."
+  :feature :cache-granular-controls
+  [card dashboard-id]
+  (when (public-settings/enable-query-caching)
+    (let [qs     [{:from   [:cache_config]
+                   :select [:id]
+                   :where  [:and [:= :model "question"]   [:= :model_id (:id card)]]}
+                  {:from   [:cache_config]
+                   :select [:id]
+                   :where  [:and [:= :model "dashboard"]  [:= :model_id dashboard-id]]}
+                  {:from   [:cache_config]
+                   :select [:id]
+                   :where  [:and [:= :model "collection"] [:= :model_id (:collection_id card)]]}
+                  {:from   [:cache_config]
+                   :select [:id]
+                   :where  [:and [:= :model "database"]   [:= :model_id (:database_id card)]]}
+                  {:from   [:cache_config]
+                   :select [:id]
+                   :where  [:= :model "root"]}]
+          q      {:select [:id]
+                  :limit  1
+                  :from   [[{:union-all qs} :unused_alias]]}
```

Comment:
Ohh, this is interesting! So I dug a little deeper and it seems like while it was not guaranteed, it was ordered (at least in Postgres) because of how execution worked. But now there is a chance to get parallel execution and so the order can change. TIL.

Reply:
done in aa80f33f24